quicklogic/qlf_k4n8/utils/create_lib.py

```diff
@@ -0,0 +1,416 @@
+                curr_str += "\n\t\t{}".format(curly_braces_close)
+                lib_data += curr_str
+            else:
+                curr_str = "\n\t\tpin ({}) {}".format(
+                    pin.name, curly_braces_open
+                )
+                cap = "0.013646"
+                max_tran = "0.2000"
+                curr_str += form_pin_header(pin.dir, cap, max_tran)
+
+                if pin.clk != '':
+                    clks = pin.clk.split(' ')
+                    for clk in clks:
+                        clk_name = clk
+                        timing_type = "rising_edge"
+                        curr_str += form_out_timing_group(
+                            clk_name, timing_type
+                        )
+                    curr_str += form_out_reset_timing_group(
+                        "RESET_N", "positive_unate", "clear"
+                    )
+                curr_str += "\n\t\t{}".format(curly_braces_close)
+                lib_data += curr_str
+        else:
+            if pin.dir == "input":
+                curr_str = "\n\t\tpin ({}) {}".format(
+                    pin.name, curly_braces_open
+                )
+                cap = "0.01686"
+                max_tran = "0.2000"
+                curr_str += form_pin_header(pin.dir, cap, max_tran)
+                curr_str += "\n\t\t{}".format(curly_braces_close)
+                lib_data += curr_str
+            else:
+                curr_str = "\n\t\tpin ({}) {}".format(
+                    pin.name, curly_braces_open
+                )
+                cap = "0.013646"
+                max_tran = "0.2000"
+                curr_str += form_pin_header(pin.dir, cap, max_tran)
+                curr_str += "\n\t\t{}".format(curly_braces_close)
+                lib_data += curr_str
+
+    dedicated_pin_lib_file = os.path.join(
+        curr_dir, '../devices/umc22/dedicated_pin_lib_data.txt'
+    )
+    dedicated_pin_data = open(dedicated_pin_lib_file, 'r').read()
+
+    inter_str = str1.replace("@dedicated_pin_data@", dedicated_pin_data)
+
+    final_str = inter_str.replace("@user_pin_data@", lib_data)
+    with open(lib_file_name, 'w') as out_fp:
+        out_fp.write(final_str)
+
+
+# =============================================================================
+
+
+def form_pin_header(direction, cap, max_tran):
+    '''
+    Form pin header section
+    '''
+    curr_str = "\n\t\t\tdirection : {};\n\t\t\tcapacitance : {};".format(
+        direction, cap
+    )
+    curr_str += "\n\t\t\tmax_transition : {};".format(max_tran)
+    return curr_str
+
+
+# =============================================================================
+
+
+def form_out_reset_timing_group(reset_name, timing_sense, timing_type):
+    '''
+    Form timing group for output pin when related pin is reset
+    '''
+    cell_fall_val = "4.6000"
+    fall_tran_val = "0.2000"
+    curly_braces_open = "{"
+    curly_braces_close = "}"
```

Comment:
You don't have to pass that as a variable to `format`. To insert a literal brace you just need to double it. `{{` will end up as `{`. Same for the closing one.

Reply:
Incorporated this in the last checkin